subsys/bluetooth/host/l2cap.c

```diff
@@ -1167,25 +1180,45 @@ static void le_ecred_conn_req(struct bt_l2cap *l2cap, uint8_t ident,
 	}
 
 response:
-	buf = l2cap_create_le_sig_pdu(buf, BT_L2CAP_ECRED_CONN_RSP, ident,
-				      sizeof(*rsp) + (sizeof(scid) * i));
+	if (result != BT_L2CAP_LE_SUCCESS) {
+		if (i == 0) {
+			/* All connections refused */
+			buf = l2cap_create_le_sig_pdu(buf, BT_L2CAP_ECRED_CONN_RSP, ident,
+				      sizeof(*rsp) + sizeof(scid));
 
-	rsp = net_buf_add(buf, sizeof(*rsp));
-	(void)memset(rsp, 0, sizeof(*rsp));
+			rsp = net_buf_add(buf, sizeof(*rsp));
+			(void)memset(rsp, 0, sizeof(*rsp));
+
+			memset(dcid, 0, sizeof(scid));
+			rsp->result = sys_cpu_to_le16(result);
+		} else {
+			buf = l2cap_create_le_sig_pdu(buf, BT_L2CAP_ECRED_CONN_RSP, ident,
+				      sizeof(*rsp) + (sizeof(scid) * i));
 
-	if (result == BT_L2CAP_LE_ERR_UNACCEPT_PARAMS ||
-	    result == BT_L2CAP_LE_ERR_PSM_NOT_SUPP ||
-	    result == BT_L2CAP_LE_ERR_AUTHENTICATION) {
-		memset(dcid, 0, sizeof(scid) * i);
+			rsp = net_buf_add(buf, sizeof(*rsp));
+			(void)memset(rsp, 0, sizeof(*rsp));
+			memset(dcid, 0, sizeof(scid) * i);
+			rsp->result = sys_cpu_to_le16(result);
+		}
 	} else if (ch) {
+		buf = l2cap_create_le_sig_pdu(buf, BT_L2CAP_ECRED_CONN_RSP, ident,
+				      sizeof(*rsp) + (sizeof(scid) * i));
+
+		rsp = net_buf_add(buf, sizeof(*rsp));
+		(void)memset(rsp, 0, sizeof(*rsp));
 		rsp->mps = sys_cpu_to_le16(ch->rx.mps);
 		rsp->mtu = sys_cpu_to_le16(ch->rx.mtu);
 		rsp->credits = sys_cpu_to_le16(ch->rx.init_credits);
+		rsp->result = sys_cpu_to_le16(result);
 	}
```

Comment:
Since there is no else to this, it means that if none of the above happens we would use buf, which is pointing to the original request to send the response. That wouldn't be a good idea I think. Is there any way we could end up in a hypothetical else here?